### PR TITLE
bpo-26510: Allow argparse add_subparsers to take `required` kwarg.

### PR DESCRIPTION
**Pull request by asottile**
I added both a test to demonstrate the current behaviour / workaround as well as the new behaviour.

I'm torn between setting the default `required=True` (which was the python2 behaviour) and setting `required=False` (the current python3 behaviour).

If we're ok with `required=True`, I'll change that up :)

https://bugs.python.org/issue26510

### CODE REVIEW

**Comment by asottile** [comment]
Actually on second thought, I'm going to try and restore the python 2 behaviour -- optional subparsers *rarely* make sense (and it'll will make writing py2 + py3 simpler).

**Review by tseaver — APPROVED** [review]
Agreed that `required=True` is the better default.

**Review by chriskuehl — APPROVED** [review]

File: Lib/argparse.py
Comment (merwok) [inline Lib/argparse.py:1071]: For maximum backward compatibility, the new parameter should be added to the end.

(I don’t personnally write calls like `something('abc', SUPPRESS, True, None, None)` but they are supported, and should not be broken when possible.)

**Review by merwok — CHANGES_REQUESTED** [review]
Please update docs and news.

File: Lib/argparse.py
Comment (asottile) [reply Lib/argparse.py:1071]: `add_subparsers` itself takes no keyword arguments so I don't think there's a concern of the private api here.

```python
>>> parser.add_subparsers('foo', True, False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: add_subparsers() takes 1 positional argument but 4 were given
```

(the 1 positional argument here being `self`)

The ordering I chose was intentional to match the signatures in the rest of the file.

Comment (merwok) [reply Lib/argparse.py:1071]: Perfect!

Comment (asottile) [reply Lib/argparse.py:1071]: @merwok do I monty-python it up now?  I'm unclear on the process for dismissing a review :P

Comment (merwok) [reply Lib/argparse.py:1071]: The review holding this up is «please update docs and news» :)

Comment (asottile) [reply Lib/argparse.py:1071]: whoops totally missed that! I'll get right on it :)

**Comment by asottile** [comment]
[I didn't expect the Spanish Inquisition!](https://youtu.be/hNavxvvN9QI)

**Review by merwok — APPROVED** [review]

**Comment by merwok** [comment]
This looks all right!  Just to be exhaustive, are the interactions ok with required options and required subparses?  (e.g. error messages mention both the missing required option and the missing required subparser)

File: Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst
Comment (merwok) [inline Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst]: @Mariatta Could you tell me if it’s the current practice to include contributors’ names in NEWS entries?

@asottile Add yourself to ACKS if you’re not already there.

Comment (Mariatta) [reply Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst]: @merwok Yes to `Patch by <contributor>` in news entries.

And also +1 to adding @asottile to Misc/ACKS.

Comment (Mariatta) [reply Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst]: `Patch by <contributor>` should end with a period. :)

Comment (asottile) [reply Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst]: I attempted to follow [this example](https://docs.python.org/devguide/committing.html#what-s-new-and-news-entries) --  if not perhaps that should be updated.

[Found myself!](https://github.com/python/cpython/blob/525269430a3f9fbb7287e4bb6b365ac216004980/Misc/ACKS#L1489)

**Comment by asottile** [comment]
> This looks all right! Just to be exhaustive, are the interactions ok with required options and required subparses? (e.g. error messages mention both the missing required option and the missing required subparser)

I can check these, though they'd already be broken (not a regression) -- I'm not actually introducing any new functionality, just a new default.  let me go through the scenarios though!

**Comment by asottile** [comment]
I fixed the only crash I could find related to this: https://github.com/python/cpython/pull/3027/commits/6e61d8a99c219ca09e070a97822201c7de814632

It was already triggerable with the following:

```python
import argparse

parser = argparse.ArgumentParser()
parsers = parser.add_subparsers()
parsers.required = True
parsers.add_parser('foo')
parsers.add_parser('bar')

parser.parse_args()
```

```
$ python3.6 test.py
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    parser.parse_args()
  File "/usr/lib/python3.6/argparse.py", line 1739, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/usr/lib/python3.6/argparse.py", line 1771, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.6/argparse.py", line 2006, in _parse_known_args
    ', '.join(required_actions))
TypeError: sequence item 0: expected str instance, NoneType found
```

The regression test I added alongside failed in the same way before patching (yay TDD).

Should this get called out separately?

File: Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst
Comment (merwok) [reply Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst]: Thanks for the help Mariatta and for the link Anthony!

**Comment by merwok** [comment]
Agreed, this looks like an unrelated problem that needs its own bug and PRs.  Good catch!

My question was more related to the interaction of a required subparser with a required option, but I guess if there is any issue it was already in 2.7 and previous versions.  I’m ready to approve this when you back out the latest commit.

**Comment by asottile** [comment]
Sweet, backed out the commit -- I'll follow up with that patch as well once this is merged!

**Comment by asottile** [comment]
Encourage those that were CC'd on this to weigh in on https://github.com/python/cpython/pull/6919 and https://bugs.python.org/issue33109

**Comment by adamjstewart** [comment]
What version of Python was this `required` parameter added to `add_subparsers`? We should add a "New in version ..." note to the docs.

**Comment by asottile** [comment]
3.7.0a2

**Comment by merwok** [comment]
The default was changed in #6919
Whatsnew was removed in #7609
Versionchanged note is being added in #16588

**Pull request merged**
